### ISSUE TITLE
HasKey should override previous HasNoKey

### ISSUE THREAD

**Issue by aquacat123** [dotnet/efcore#24789]
I have a view that returns a key column (using ROW_NUMBER() OVER in SQL as TempID), however using db-scaffold the generated code for the view contains "entity.HasNoKey();".  This is causing queries to break because the linq subqueries are requiring the parent projection to contain a primary key.  Manually changing the generated code will allow my queries to work, but is a problem because as soon as i regenerate my context using db-scaffold then the custom change is lost.  If i 

There has to be a way in EF Core 5 to specify a key column in a view (without altering generated code). 

I tried using the "OnModelCreatingPartial" method to set the key but this causes a "The key cannot be added to keyless type" exception.

        partial void OnModelCreatingPartial(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<TemplateLocationView>().HasKey(e => e.TempId);
}

**Comment by slavanap** [comment]
As for the key,

```c#
partial void OnModelCreatingPartial(ModelBuilder modelBuilder)
{
    modelBuilder.Entity<TemplateLocationView>(entity => {
        entity.Metadata.IsKeyless = false;
        entity.HasKey(e => e.TempId);
    });
}
```

shall work.
@aquacat123

**Comment by xutbest** [comment]
I have the same issue in .NET 6.0? Is there a new way to fix it?

**Comment by roji** [comment]
@xutbest this issue was fixed for EF Core 6.0 - are you sure you're EF Core 6.0 or 7.0? If you are, please open a new issue with a minimal, runnable code sample that shows the problem.
